syntax = "proto3";

package multiplication;

message Point {
  int32 number = 1;
}

message Multiplier {
  Point number = 1;
}

message Rectangle {
  Point lo = 1;
  Point hi = 2;
}

message MultiplicationSummary {
  int32 point_count = 1;
  int32 feature_count = 2;
  int64 multiplication_result = 3;
  int32 elapsed_time = 4;
}

service Multiplication {
  rpc GetMultiplier(Point) returns (Multiplier);
  rpc ListMultipliers(Rectangle) returns (stream Multiplier);
  rpc RecordMultiplication(stream Point) returns (MultiplicationSummary);
  rpc MultiplicationChat(stream Multiplier) returns (stream Multiplier);
}
